src/eg/src/guardian_share.rs

```diff
@@ -226,14 +290,87 @@ impl GuardianEncryptedShare {
             return Err(DecryptionError::InvalidMAC);
         }
 
+        // Decryption as in Equation `20`
         let p_l_bytes = xor(self.c1.0.as_slice(), k1.0.as_slice());
+        let p_l = BigUint::from_bytes_be(p_l_bytes.as_slice());
+
+        // Share validity check
+        let g_p_l = fixed_parameters.g.modpow(&p_l, p);
+        // RHS of Equation `21`
+        let l = &BigUint::from(l);
+        let vec_k_i_j = &dealer_public_key.coefficient_commitments.0;
+        let rhs = vec_k_i_j
+            .iter()
+            .enumerate()
+            .fold(BigUint::one(), |prod, (j, k_i_j)| {
+                //This is fine as j < k
+                #[allow(clippy::unwrap_used)]
+                let j: u32 = j.try_into().unwrap();
+                (prod * k_i_j.0.modpow(&l.pow(j), p)) % p
+            });
+        if g_p_l != rhs {
+            return Err(DecryptionError::InvalidShare);
+        }
+
+        Ok(p_l)
+    }
+
+    /// This function validates a [`GuardianEncryptedShare`] with respect to given [`GuardianEncryptionSecret`] and [`GuardianPublicKey`] of dealer and recipient in case of a dispute.
```

Comment:
Could we add more into the documentation about the output of the function? E.g., something along the lines of "This function returns false if the dealer's published secrets do not match the published ciphertext, or the equation `21`. If they match, the function returns true, meaning the recipient's claim about the wrongdoing by the dealer is dismissed."

Reply:
I have added your suggestion.